### PR TITLE
Use Node IP Address instead of Node.Name in minion.ResourceLocation.

### PR DESCRIPTION
**Pull request by cjcullen**
Refactor GetNodeHostIP into pkg/util/node (instead of pkg/util to break import cycle).

Include internalIP in gce NodeAddresses. Remove NodeLegacyHostIP

### CODE REVIEW

**Comment by cjcullen** [comment]
This should fix the last of the "name as address" usages for #2462

File: pkg/cloudprovider/gce/gce.go [hunk 1]
```diff
@@ -480,11 +482,14 @@ func (gce *GCECloud) NodeAddresses(_ string) ([]api.NodeAddress, error) {
 	if err != nil {
 		return nil, fmt.Errorf("couldn't get external IP: %v", err)
 	}
+	internalIP, err := gce.metadataAccess(INTERNAL_IP_METADATA_URL)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get external IP: %v", err)
```
Comment (bgrant0607) [inline pkg/cloudprovider/gce/gce.go]: internal IP

File: pkg/util/node/node.go [hunk 2]
```diff
@@ -34,3 +37,25 @@ func GetHostname(hostnameOverride string) string {
...
+// 1. NodeLegacyHostIP
+// 2. NodeInternalIP
+// 3. NodeExternalIP
+func GetNodeHostIP(node *api.Node) (net.IP, error) {
+	addresses := node.Status.Addresses
+	addressMap := make(map[api.NodeAddressType][]api.NodeAddress)
+	for i := range addresses {
+		addressMap[addresses[i].Type] = append(addressMap[addresses[i].Type], addresses[i])
+	}
+	if addresses, ok := addressMap[api.NodeLegacyHostIP]; ok {
```
Comment (bgrant0607) [inline pkg/util/node/node.go]: Please remove LegacyHostIP. We're in the process of nuking it, anyway.

File: pkg/cloudprovider/gce/gce.go [hunk 3]
```diff
@@ -480,11 +482,14 @@ func (gce *GCECloud) NodeAddresses(_ string) ([]api.NodeAddress, error) {
...
+	}
 
 	return []api.NodeAddress{
 		{Type: api.NodeExternalIP, Address: externalIP},
-		// TODO(mbforbes): Remove NodeLegacyHostIP once v1beta1 is removed.
-		{Type: api.NodeLegacyHostIP, Address: externalIP},
+		{Type: api.NodeInternalIP, Address: internalIP},
```
Comment (bgrant0607) [inline pkg/cloudprovider/gce/gce.go]: For clarity, I'd put these in the priority order expected by GetNodeHostIP: InternalIP should go first.

File: pkg/cloudprovider/gce/gce.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/cloudprovider/gce/gce.go]: And I'd put this above the externalIP code, to be consistent with the ordering everywhere else.

File: pkg/cloudprovider/gce/gce.go (hunk 3 above)
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: Done.

File: pkg/util/node/node.go (hunk 2 above)
Comment (cjcullen) [inline pkg/util/node/node.go]: Done.

File: pkg/cloudprovider/gce/gce.go (hunk 1 above)
Comment (cjcullen) [inline pkg/cloudprovider/gce/gce.go]: Done.

**Comment by cjcullen** [comment]
First round of comments addressed. PTAL.

**Comment by bgrant0607** [comment]
Thanks. LGTM. Did you run e2e?

**Comment by cjcullen** [comment]
Just finished:
Summarizing 1 Failure:

[Fail] Cluster level logging using Elasticsearch [It] should check that logs from pods on all nodes are ingested into Elasticsearch 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:210

Ran 61 of 70 Specs in 3281.773 seconds
FAIL! -- 60 Passed | 1 Failed | 2 Pending | 7 Skipped --- FAIL: TestE2E (3392.29s)
FAIL

**Comment by bgrant0607** [comment]
Thanks. I think #9148 has a fix for that.

**Comment by karlkfi** [comment]
I don't have a lot of context on this change, but I'm getting an error now: 
`kubelet.go:2099] Cannot get host IP: host IP unknown; known addresses: [{LegacyHostIP 172.17.13.122}]`

That's a kublet saying it can't find it's own host, even tho it lists it as a known address. I see the `GetNodeHostIP` added in this PR only checks for addresses of type `InternalIP` and `ExternalIP`. Does that mean my configuration is out of date and I need to specify internal/external IPs somehow?

**Comment by bgrant0607** [comment]
What cloudprovider?
We'll try to get in the changes that replace LegacyHostIP ASAP. In the meantime, it could search all 3.
cc @caesarxuchao

**Comment by bgrant0607** [comment]
cc @krousey

**Comment by karlkfi** [comment]
Mesos cloud provider. I just noticed that it's only populating the old one: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/cloudprovider/mesos/mesos.go#L240

**Comment by bgrant0607** [comment]
@jdef Should Mesos's LegacyHostIP be replaced with InternalIP or ExternalIP?

**Comment by bcbroussard** [comment]
@karlkfi @bgrant0607 I'm also seeing this when proxying to nodes for cAdvisor requests. I'm running on a local cluster. `http://localhost:8080/api/v1beta3/proxy/nodes/127.0.0.1:4194/api/v1.0/containers/`

**Comment by bgrant0607** [comment]
Rollback PR is #9232.

**Comment by karlkfi** [comment]
I like the idea of falling back to NodeLegacyHostIP if NodeExternalIP/NodeInternalIP aren't present. That's what's happening in the e2e service tests: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/service.go#L979

**Comment by jdef** [comment]
If LegacyHostIP is going away, then .. InternalIP seems more correct. The
mesos cloud provider attempts to take a pretty generic view of the network
since it can't really make many assumptions about the actual cloud that
you're running on. We want to circle back on this at some point to build in
some kind of pluggable/wrapper support for existing cloud providers,
leveraging the work that others have already done to get k8s running on
their cloud.

**Comment by bgrant0607** [comment]
See #9267

**Pull request merged**